8. Stateless

iphydf opened the pull request:
**What are we doing?**

We are moving towards stateless callbacks. This means that when registering a
callback, you no longer pass a user data pointer. Instead, you pass a user data
pointer to tox_iterate. This pointer is threaded through the code, passed to
each callback. The callback can modify the data pointed at. An extra indirection
will be needed if the pointer itself can change.

**Why?**

Currently, callbacks are registered with a user data pointer. This means the
library has N pointers for N different callbacks. These pointers need to be
managed by the client code. Managing the lifetime of the pointee can be
difficult. In C++, it takes special effort to ensure that the lifetime of user
data extends at least beyond the lifetime of the Tox instance. For other
languages, the situation is much worse. Java and other garbage collected
languages may move objects in memory, so the pointers are not stable. Tox4j goes
through a lot of effort to make the Java/Scala user experience a pleasant one by
keeping a global array of Tox+userdata on the C++ side, and communicating via
protobufs. A Haskell FFI would have to do similarly complex tricks.

Stateless callbacks ensure that a user data pointer only needs to live during a
single function call. This means that the user code (or language runtime) can
move the data around at will, as long as it sets the new location in the
callback.

**How?**

We are doing this change one callback at a time. After each callback, we ensure
that everything still works as expected. This means the toxcore change will
require 15 Pull Requests.

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/toktok/toxcore/27)
<!-- Reviewable:end -->

nurupo commented:
Very confused with the `06bebba` commit. Are you going to remove `userdata` from all of the callbacks? Why?

tux3 commented:
Review status: 0 of 13 files reviewed at latest revision, 1 unresolved discussion.

---

*[toxcore/tox.h, line 765 \[r1\]](https://reviewable.io:443/reviews/toktok/toxcore/27#-KPJgYti2G8BPrJo6jN2:-KPJgYtjniQ1_EHuJ1v8:b6211lu) ([raw file](https://github.com/toktok/toxcore/blob/a697c3c6ce2d35d68dc65c5c0e103d3662833e98/toxcore/tox.h#L765)):*
> ```objc
>  * @param connection_status Whether we are connected to the DHT.
>  */
> typedef void *tox_self_connection_status_cb(Tox *tox, TOX_CONNECTION connection_status, void *user_data);
> ```

A comment to document what may/should be returned by the callback would be good.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/27)*
<!-- Sent from Reviewable.io -->

tux3 commented:
Reviewed 13 of 13 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/27)*
<!-- Sent from Reviewable.io -->

iphydf commented:
Review status: 11 of 14 files reviewed at latest revision, 1 unresolved discussion.

---

*[toxcore/tox.h, line 765 \[r1\]](https://reviewable.io:443/reviews/toktok/toxcore/27#-KPJgYti2G8BPrJo6jN2:-KPK4IGmuibGbuH-Ll8o:b-aw7arl) ([raw file](https://github.com/toktok/toxcore/blob/a697c3c6ce2d35d68dc65c5c0e103d3662833e98/toxcore/tox.h#L765)):*
<details><summary><i>Previously, tux3 (Tux3) wrote…</i></summary>
> A comment to document what may/should be returned by the callback would be good.
</details>
Added some documentation to the top. PTAL.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/27)*
<!-- Sent from Reviewable.io -->

iphydf commented:
Added information to the commit message. PTAL.

---

Review status: 11 of 14 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/27#-:-KPKBWodiRQhp8Nym7ek:b-srn6fg)*
<!-- Sent from Reviewable.io -->

tux3 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/27#-:-KPKDFOkwZ1Aci9guGRs:bnfp4nl)*
<!-- Sent from Reviewable.io -->

nurupo commented:
Reviewed 11 of 13 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[other/apidsl/tox.in.h, line 754 \[r2\]](https://reviewable.io:443/reviews/toktok/toxcore/27#-KPKRS2Myj_VOkvZZKpa:-KPKRS2Myj_VOkvZZKpb:b70ouyf) ([raw file](https://github.com/toktok/toxcore/blob/ca8e9269802d06268eadca4fd0b7b123eb19b806/other/apidsl/tox.in.h#L754)):*
> ```objc
>  * milliseconds.
>  */
> `a iterate(`a user_data);
> ```

I was totally blown away by the cyclone here.

I suggest having "boxed types" starting with a capital letter à la Java: Void -> void*, Int -> int*, Uint32_t -> uint32_t*, etc.

While most of apidsl is C++ class-y/namespace-y and intuitive, this is straight out weird.

---

*[other/apidsl/tox.in.h, line 1242 \[r2\]](https://reviewable.io:443/reviews/toktok/toxcore/27#-KPKPYgxAdHAiwC82Aiq:-KPKPYgyDyuK7uvc-0O-:b-e8uisc) ([raw file](https://github.com/toktok/toxcore/blob/ca8e9269802d06268eadca4fd0b7b123eb19b806/other/apidsl/tox.in.h#L1242)):*
>and store the name

What name? This callback doesn't deal with any names.

---

*[other/apidsl/tox.in.h, line 1274 \[r2\]](https://reviewable.io:443/reviews/toktok/toxcore/27#-KPKPo447V8c0GbBgaFv:-KPKPo447V8c0GbBgaFw:b-e8uisc) ([raw file](https://github.com/toktok/toxcore/blob/ca8e9269802d06268eadca4fd0b7b123eb19b806/other/apidsl/tox.in.h#L1274)):*
>and store the name

What name? This callback doesn't deal with any names.

---

*[other/apidsl/tox.in.h, line 1307 \[r2\]](https://reviewable.io:443/reviews/toktok/toxcore/27#-KPKPqnRSq5fPP50ACv6:-KPKPqnRSq5fPP50ACv7:b-e8uisc) ([raw file](https://github.com/toktok/toxcore/blob/ca8e9269802d06268eadca4fd0b7b123eb19b806/other/apidsl/tox.in.h#L1307)):*
>and store the name

What name? This callback doesn't deal with any names.

---

*[other/apidsl/tox.in.h, line 1342 \[r2\]](https://reviewable.io:443/reviews/toktok/toxcore/27#-KPKPsNk0N3ksN0H1OKs:-KPKPsNk0N3ksN0H1OKt:b-e8uisc) ([raw file](https://github.com/toktok/toxcore/blob/ca8e9269802d06268eadca4fd0b7b123eb19b806/other/apidsl/tox.in.h#L1342)):*
>and store the name

What name? This callback doesn't deal with any names.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/27)*
<!-- Sent from Reviewable.io -->

iphydf commented:
Review status: 13 of 14 files reviewed at latest revision, 5 unresolved discussions.

---

*[other/apidsl/tox.in.h, line 754 \[r2\]](https://reviewable.io:443/reviews/toktok/toxcore/27#-KPKRS2Myj_VOkvZZKpa:-KPKTsvc2gqPO8G1Hl7m:b-10e2w3) ([raw file](https://github.com/toktok/toxcore/blob/ca8e9269802d06268eadca4fd0b7b123eb19b806/other/apidsl/tox.in.h#L754)):*
<details><summary><i>Previously, nurupo wrote…</i></summary>
> I was totally blown away by the cyclone here.
> 
> I suggest having "boxed types" starting with a capital letter à la Java: Void -> void*, Int -> int*, Uint32_t -> uint32_t*, etc.
> 
> While most of apidsl is C++ class-y/namespace-y and intuitive, this is straight out weird.
</details>
What notation would you suggest for this? The meaning is like `template<typename T> T iterate(T user_data);`

---

*[other/apidsl/tox.in.h, line 1242 \[r2\]](https://reviewable.io:443/reviews/toktok/toxcore/27#-KPKPYgxAdHAiwC82Aiq:-KPKTpfBq0R19oAE1QaZ:b-896fix) ([raw file](https://github.com/toktok/toxcore/blob/ca8e9269802d06268eadca4fd0b7b123eb19b806/other/apidsl/tox.in.h#L1242)):*
<details><summary><i>Previously, nurupo wrote…</i></summary>
> >and store the name
> 
> What name? This callback doesn't deal with any names.
</details>
Done.

---

*[other/apidsl/tox.in.h, line 1274 \[r2\]](https://reviewable.io:443/reviews/toktok/toxcore/27#-KPKPo447V8c0GbBgaFv:-KPKTqc2akugIYQbTh8U:b-896fix) ([raw file](https://github.com/toktok/toxcore/blob/ca8e9269802d06268eadca4fd0b7b123eb19b806/other/apidsl/tox.in.h#L1274)):*
<details><summary><i>Previously, nurupo wrote…</i></summary>
> >and store the name
> 
> What name? This callback doesn't deal with any names.
</details>
Done.

---

*[other/apidsl/tox.in.h, line 1307 \[r2\]](https://reviewable.io:443/reviews/toktok/toxcore/27#-KPKPqnRSq5fPP50ACv6:-KPKTrOHebXsgMYP4wY0:b-896fix) ([raw file](https://github.com/toktok/toxcore/blob/ca8e9269802d06268eadca4fd0b7b123eb19b806/other/apidsl/tox.in.h#L1307)):*
<details><summary><i>Previously, nurupo wrote…</i></summary>
> >and store the name
> 
> What name? This callback doesn't deal with any names.
</details>
Done.

---

*[other/apidsl/tox.in.h, line 1342 \[r2\]](https://reviewable.io:443/reviews/toktok/toxcore/27#-KPKPsNk0N3ksN0H1OKs:-KPKTsBcU_iyp2R2zzlo:b-896fix) ([raw file](https://github.com/toktok/toxcore/blob/ca8e9269802d06268eadca4fd0b7b123eb19b806/other/apidsl/tox.in.h#L1342)):*
<details><summary><i>Previously, nurupo wrote…</i></summary>
> >and store the name
> 
> What name? This callback doesn't deal with any names.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/27)*
<!-- Sent from Reviewable.io -->

nurupo commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[other/apidsl/tox.in.h, line 754 \[r2\]](https://reviewable.io:443/reviews/toktok/toxcore/27#-KPKRS2Myj_VOkvZZKpa:-KPKUMQnbpwIFKhoCvfA:b-7p5vu1) ([raw file](https://github.com/toktok/toxcore/blob/ca8e9269802d06268eadca4fd0b7b123eb19b806/other/apidsl/tox.in.h#L754)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> What notation would you suggest for this? The meaning is like `template<typename T> T iterate(T user_data);`
</details>
Uh, markdown has eaten my stars and converted them into italics. `Void` -> `void*`, `Int` -> `int*`, `Uint32_t` -> `uint32_t*`.

Template syntax is fine with me, but why not just use `void*` or similar?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/27)*
<!-- Sent from Reviewable.io -->

nurupo commented:
Review status: 13 of 14 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[other/apidsl/tox.in.h, line 1242 \[r2\]](https://reviewable.io:443/reviews/toktok/toxcore/27#-KPKPYgxAdHAiwC82Aiq:-KPKXOrzhSpLy0irpSsC:b-uy0hji) ([raw file](https://github.com/toktok/toxcore/blob/ca8e9269802d06268eadca4fd0b7b123eb19b806/other/apidsl/tox.in.h#L1242)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> Done.
</details>
Now run apidsl to apply these changes to `tox.h`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/27)*
<!-- Sent from Reviewable.io -->

GrayHatter commented:
Reviewed 11 of 13 files at r1, 1 of 3 files at r2.
Review status: 13 of 14 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[toxcore/Messenger.h, line 279 \[r3\]](https://reviewable.io:443/reviews/toktok/toxcore/27#-KPKez34YmSWAV271aTr:-KPKez35k_s67cS98MCD:b-q78nze) ([raw file](https://github.com/toktok/toxcore/blob/70d64619039d808ded122330f15715c9b8b6de57/toxcore/Messenger.h#L279)):*
> ```objc
>     void *lossless_packethandler_userdata;
> 
>     void *(*core_connection_change)(struct Messenger *m, unsigned int, void *);
> ```

We should change this from an enumerated type, to it's current typedef' in tox.h

---

*[toxcore/tox.h, line 1207 \[r3\]](https://reviewable.io:443/reviews/toktok/toxcore/27#-KPKZrWa24OUfQkArPda:-KPKZrWbPKfL-s0qeCy-:bx36lcc) ([raw file](https://github.com/toktok/toxcore/blob/70d64619039d808ded122330f15715c9b8b6de57/toxcore/tox.h#L1207)):*
> ```objc
>      * No friend with the given number exists on the friend list.
>      *
>      * @deprecated The library may not store this information in the future.
> ```

Are we deprecating just this error return, or this whole fxn as well?

If it's the latter, we need to move this comment down as well.

---

*[toxcore/tox.h, line 1349 \[r3\]](https://reviewable.io:443/reviews/toktok/toxcore/27#-KPKaDrM4P_OCIdWj8Cu:-KPKaDrM4P_OCIdWj8Cv:b-1fh8uz) ([raw file](https://github.com/toktok/toxcore/blob/70d64619039d808ded122330f15715c9b8b6de57/toxcore/tox.h#L1349)):*
> ```objc
>  * `friend_status` callback.
>  *
>  * @deprecated Use the friend status callback and store the name in user code.
> ```

store the name? (x3)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/27)*
<!-- Sent from Reviewable.io -->

iphydf commented:
Review status: 12 of 14 files reviewed at latest revision, 5 unresolved discussions.

---

*[other/apidsl/tox.in.h, line 754 \[r2\]](https://reviewable.io:443/reviews/toktok/toxcore/27#-KPKRS2Myj_VOkvZZKpa:-KPMXtv4Gjvsn_6HxZQE:b3j2k3e) ([raw file](https://github.com/toktok/toxcore/blob/ca8e9269802d06268eadca4fd0b7b123eb19b806/other/apidsl/tox.in.h#L754)):*
<details><summary><i>Previously, nurupo wrote…</i></summary>
> Uh, markdown has eaten my stars and converted them into italics. `Void` -> `void*`, `Int` -> `int*`, `Uint32_t` -> `uint32_t*`.
> 
> Template syntax is fine with me, but why not just use `void*` or similar?
</details>
How about this? "any" is "void *". Apidsl intentionally does not support arbitrary pointers, so there needs to be special support for void* and const char* (= string).

---

*[other/apidsl/tox.in.h, line 1242 \[r2\]](https://reviewable.io:443/reviews/toktok/toxcore/27#-KPKPYgxAdHAiwC82Aiq:-KPMY6Ijbd3P85yl0SvW:b-896fix) ([raw file](https://github.com/toktok/toxcore/blob/ca8e9269802d06268eadca4fd0b7b123eb19b806/other/apidsl/tox.in.h#L1242)):*
<details><summary><i>Previously, nurupo wrote…</i></summary>
> Now run apidsl to apply these changes to `tox.h`.
</details>
Done.

---

*[toxcore/Messenger.h, line 279 \[r3\]](https://reviewable.io:443/reviews/toktok/toxcore/27#-KPKez34YmSWAV271aTr:-KPMYCxbxip4s8OBlzkr:bvg0sge) ([raw file](https://github.com/toktok/toxcore/blob/70d64619039d808ded122330f15715c9b8b6de57/toxcore/Messenger.h#L279)):*
<details><summary><i>Previously, GrayHatter (Gregory Mullen) wrote…</i></summary>
> We should change this from an enumerated type, to it's current typedef' in tox.h
</details>
Yeah. Not in scope here.

---

*[toxcore/tox.h, line 1207 \[r3\]](https://reviewable.io:443/reviews/toktok/toxcore/27#-KPKZrWa24OUfQkArPda:-KPMW0m23xd7sPob_ucV:b1ghm00) ([raw file](https://github.com/toktok/toxcore/blob/70d64619039d808ded122330f15715c9b8b6de57/toxcore/tox.h#L1207)):*
<details><summary><i>Previously, GrayHatter (Gregory Mullen) wrote…</i></summary>
> Are we deprecating just this error return, or this whole fxn as well?
> 
> If it's the latter, we need to move this comment down as well.
</details>
Removed deprecation for now.

---

*[toxcore/tox.h, line 1349 \[r3\]](https://reviewable.io:443/reviews/toktok/toxcore/27#-KPKaDrM4P_OCIdWj8Cu:-KPMW4b0lDizWDuF9UBV:b-896fix) ([raw file](https://github.com/toktok/toxcore/blob/70d64619039d808ded122330f15715c9b8b6de57/toxcore/tox.h#L1349)):*
<details><summary><i>Previously, GrayHatter (Gregory Mullen) wrote…</i></summary>
> store the name? (x3)
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/27)*
<!-- Sent from Reviewable.io -->

GrayHatter commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[other/apidsl/tox.in.h, line 754 \[r2\]](https://reviewable.io:443/reviews/toktok/toxcore/27#-KPKRS2Myj_VOkvZZKpa:-KPMYhQFhwBOQyQkQnFt:b-rwsjw9) ([raw file](https://github.com/toktok/toxcore/blob/ca8e9269802d06268eadca4fd0b7b123eb19b806/other/apidsl/tox.in.h#L754)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> How about this? "any" is "void *". Apidsl intentionally does not support arbitrary pointers, so there needs to be special support for void* and const char* (= string).
</details>
I like `any`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/27#-:-KPMYs53gXVq3xzTccNd:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tux3 commented:
I like this rev much more than the last, <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Reviewed 6 of 6 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/27#-:-KPNIqE5SEuDo8Fp02Zu:bpgje4m)*
<!-- Sent from Reviewable.io -->

JFreegman commented:
Reviewed 7 of 13 files at r1, 1 of 3 files at r2, 6 of 6 files at r5, 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks broke.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/27)*
<!-- Sent from Reviewable.io -->

JFreegman commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/27#-:-KPNzaB9KynZP7MFkttn:bnfp4nl)*
<!-- Sent from Reviewable.io -->

iphydf commented:
I'm going to merge this PR now, since everybody agreed with the change.
